### ISSUE TITLE
Test Results of optimizekel functionality with code from 2019-11-06 in branch openNCA-2.0.5-M1-Release

### ISSUE THREAD

**Issue by tensfeldt** [tensfeldt/openNCA#2]
# <span class="timestamp-wrapper"><span class="timestamp">&lt;2019-11-07 Thu&gt; </span></span> openNCA-2.0.5-M1-Release development branch testing

## tc010 M1SD -

ran with both mct$optimizekel = 0 & 1, ran as expected in both cases  
confirmation of whether the optimization worked completely as expected was not undertaken but will be evaluated next  
need to confirm whether changes to MCT$FLGACCEPTKELCRIT are respected as well  

## tc011 M1SS - errors with unit conversion optimizekel==0 (non optimize)

these errors were confirmed to be pre-existing in the master code branch  

### > results<sub>list</sub> <- run<sub>computation</sub>(data=d, map=mct, flag=flags, parameterset=parameterset)

Error in withCallingHandlers(expr, warning = function(w) invokeRestart("muffleWarning")) :  
  (list) object cannot be coerced to type 'double'  
In addition: There were 50 or more warnings (use warnings() to see the first 50)  
> warnings()  
Warning messages:  
1: In run<sub>M1</sub><sub>SS</sub><sub>computation</sub>(data = merged<sub>data</sub>, map = map<sub>data</sub>,  &#x2026; :  
  Flag 'FLGEMESIS' is not present in the dataset  
2: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
3: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
4: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
5: In matrix(value, n, p) :  
  data length [157] is not a sub-multiple or multiple of the number of columns [156]  
6: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
7: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
8: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
9: In matrix(value, n, p) :  
  data length [157] is not a sub-multiple or multiple of the number of columns [156]  
10: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
11: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
12: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
13: In matrix(value, n, p) :  
  data length [157] is not a sub-multiple or multiple of the number of columns [156]  
14: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
15: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
16: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
17: In matrix(value, n, p) :  
  data length [157] is not a sub-multiple or multiple of the number of columns [156]  
18: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
19: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
20: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
21: In matrix(value, n, p) :  
  data length [157] is not a sub-multiple or multiple of the number of columns [156]  
22: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
23: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
24: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
25: In matrix(value, n, p) :  
  data length [157] is not a sub-multiple or multiple of the number of columns [156]  
26: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
27: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
28: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
29: In matrix(value, n, p) :  
  data length [157] is not a sub-multiple or multiple of the number of columns [156]  
30: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
31: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
32: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
33: In matrix(value, n, p) :  
  data length [157] is not a sub-multiple or multiple of the number of columns [156]  
34: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
35: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
36: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
37: In matrix(value, n, p) :  
  data length [157] is not a sub-multiple or multiple of the number of columns [156]  
38: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
39: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
40: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
41: In matrix(value, n, p) :  
  data length [157] is not a sub-multiple or multiple of the number of columns [156]  
42: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
43: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
44: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
45: In matrix(value, n, p) :  
  data length [157] is not a sub-multiple or multiple of the number of columns [156]  
46: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
47: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
48: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  
49: In matrix(value, n, p) :  
  data length [157] is not a sub-multiple or multiple of the number of columns [156]  
50: In unit<sub>conversion</sub>(data = data<sub>data</sub>, map = map<sub>data</sub>, result = data<sub>data</sub>,  &#x2026; :  
  'CONCOUTPUTUNIT' is not present in the proper form! Please try again using 'Amount/Volume' format!  

### run 3 - flags$FLGEXKEL <- rep(1, nrow(flags)) # i.e. just didn't select ANY data points for inclusion in terminal elimination regression

to test further  
and no regression was performed apparently when executing run<sub>computation</sub>, i.e.  

all values of KEL were NA  

# tc013 M2SD - took 2:10.20 min to execute

**Comment by jhhughes256** [comment]
### Update

As of commit a06c1c7461 the issue with tc012 still remains. It can also be replicated using tc014, tc017 & tc018. This issue does not occur when using tc010, tc013, tc015, tc016.

It appears that OPTIMIZEKEL in the mct does not turn on optimisation for these cases. Instead it appears to read FLGEXKEL in the flags data, as when these values are set to 1 (exclude these points for KEL if OPTIMIZEKEL == 0) output for KEL is NA.

Additionally tc013 and tc015 still take a significant amount of time to run (1:40 & 4:58 mins respectively in my testing environment). This may be better covered in a separate issue.

### Output for tc012

`run_computation` output `data_out` for tc012 shown below, with reproducible example included to replicate the final output.

#### mct$OPTIMIZEKEL == 0 and unchanged flags file

```
         KEL KELTMLO KELTMHI KELNOPT    KELRSQ   KELRSQA FLGACCEPTKEL
1 0.10416094      18      24       3 0.9995933 0.9991866        FALSE
2 0.12611367      12      24       3 0.9999997 0.9999995        FALSE
3 0.11779122      12      24       3 0.9970626 0.9941252        FALSE
4 0.09888024      16      24       4 0.9990935 0.9986402        FALSE
5 0.10647171      12      24       3 0.9975960 0.9951920        FALSE
6 0.13200860      16      24       4 0.9992316 0.9988474        FALSE
```

#### mct$OPTIMIZEKEL == 1 and unchanged flags file

```
         KEL KELTMLO KELTMHI KELNOPT    KELRSQ   KELRSQA FLGACCEPTKEL
1 0.10416094      18      24       3 0.9995933 0.9991866        FALSE
2 0.12611367      12      24       3 0.9999997 0.9999995        FALSE
3 0.11779122      12      24       3 0.9970626 0.9941252        FALSE
4 0.09888024      16      24       4 0.9990935 0.9986402        FALSE
5 0.10647171      12      24       3 0.9975960 0.9951920        FALSE
6 0.13200860      16      24       4 0.9992316 0.9988474        FALSE
```

#### mct$OPTIMIZEKEL == 0 and flags$FLGEXKEL == 1

```
  KEL KELTMLO KELTMHI KELNOPT KELRSQ KELRSQA FLGACCEPTKEL
1  NA      NA      NA       0     NA      NA        FALSE
2  NA      NA      NA       0     NA      NA        FALSE
3  NA      NA      NA       0     NA      NA        FALSE
4  NA      NA      NA       0     NA      NA        FALSE
5  NA      NA      NA       0     NA      NA        FALSE
6  NA      NA      NA       0     NA      NA        FALSE
```

#### mct$OPTIMIZEKEL == 1 and flags$FLGEXKEL == 1

```
  KEL KELTMLO KELTMHI KELNOPT KELRSQ KELRSQA FLGACCEPTKEL
1  NA      NA      NA       0     NA      NA        FALSE
2  NA      NA      NA       0     NA      NA        FALSE
3  NA      NA      NA       0     NA      NA        FALSE
4  NA      NA      NA       0     NA      NA        FALSE
5  NA      NA      NA       0     NA      NA        FALSE
6  NA      NA      NA       0     NA      NA        FALSE
```

### Reproducible Example

```
#### Load Libraries ------------------------------------------------------ ####
library(readr)
library(jsonlite)
library(testthat)

#### Load specific code-base for OpenNCA --------------------------------- ####
# pkgload::load_all(path = "codebase/openNCA-master/openNCA")
# pkgload::load_all(path = "codebase/openNCA-b1977923fc/openNCA")  # 11/8-v1
# pkgload::load_all(path = "codebase/openNCA-051805c887/openNCA")  # 11/8-v2
# pkgload::load_all(path = "codebase/openNCA-1fc479c2ca/openNCA")  # 11/11
pkgload::load_all(path = "codebase/openNCA-a06c1c7461/openNCA")  # 11/12

#### Clean test environment ---------------------------------------------- ####
# Clean environment
rm(params, conc, vlist, results, results_list, r, e, mct, d, flags, p, klist, n, 
  ninc, nexc, k, rlist, dlist, nlist, slist, df, xmin, xmax, dk, flag_data, 
  est_data, outputdir)

profileid <- "SDEID"
args = commandArgs(trailingOnly=TRUE)
params = c()

#### Test Case 012 ------------------------------------------------------- ####
# M1 SS
# p <- "C:/Users/tensfeldt/Box Sync/EQuIP Computation Engine/Rudraya_executed_test_cases/tc012"
testcase.id <- "tc012"
p <- "testcase/tc012_M1SS"
params["INPUT1"] = file.path(p, "A0081216.in.csv")
params["parameterset"] ="PARAMETERLIST"
params["map"] = file.path(p, "tc012_mct_updated.csv")
params["flags"] = file.path(p, "A0081216_flag_data.in.csv")
params["OUTPUT"] ="."
print(params)

mct <- read_csv(params["map"])
write_file(toJSON(mct, na="string"), path=file.path(p, "tc012_mct_updated.json"))
params["map"] = file.path(p, "tc012_mct_updated.json")

flags <- read_csv(params["flags"])
write_file(toJSON(flags, na="string"), path=file.path(p, "tc012_flags.json"))
params["flags"] = file.path(p, "tc012_flags.json")

#### Read full concentration dataset ------------------------------------- ####
# Read conc
conc <- read.csv(file=params["INPUT1"], header=TRUE, sep=",", as.is=TRUE)
vlist <- c("PKDATAROWID", "PKPTMS", "PKCNCN", "PERIOD", "TREATXT")

# Read mct
mct <- params["map"]
###mct <- read_file(mct)
###mct <- gsub("[\\]+", "", mct)
mct <- fromJSON(mct)

# Read flags
flags <- params["flags"]
###flags <- read_file(flags)
###flags <- gsub("[\\]+", "", flags)
flags <- fromJSON(flags)

# ParameterSet
if(is.na(params["parameterset"])) { params["parameterset"] <- "PARAMETERDISPLAYLIST" }
parameterset <- params["parameterset"]

# FLAGS is a subset for the specific profile, subset concentration dataframe 
#  to those records by PKDATAROWID
d <- conc[is.element(conc$PKDATAROWID, flags$PKDATAROWID),]

# Add in missing required parameters for virtual compute call
klist <- c("KEL", "KELRSQ", "KELNOPT", "THALFF")
if(!is.element("PARAMETERDISPLAYLIST", names(mct))) { mct$PARAMETERDISPLAYLIST <- "" }
for(i in klist) {
  ###if(regexpr(i, mct$PARAMETERDISPLAYLIST, ignore.case=TRUE)==-1) { mct$PARAMETERDISPLAYLIST <- c(mct$PARAMETERDISPLAYLIST, i) }
  if(regexpr(i, mct$PARAMETERDISPLAYLIST, ignore.case=TRUE)==-1) { mct$PARAMETERDISPLAYLIST <- paste(mct$PARAMETERDISPLAYLIST, i, sep=";") }
}

# Remove any leading semi-colons ";"
mct$PARAMETERDISPLAYLIST <- gsub("^;*", "", mct$PARAMETERDISPLAYLIST, ignore.case=TRUE, perl=TRUE)

# For virtual compute call only: Subset the parameters to be calculated to those in parameterdisplaylist
if(!is.element("PARAMETERLIST", names(mct))) { mct$PARAMETERLIST <- "" }

# Check whether TIME selection has been set in MCT - default to NOMINAL
if(!is.element("TIME", names(mct))) { mct$TIME <- "NOMINAL" }

# Determine whether there are any points selected for inclusion in the terminal elimination phase via 
#  FLGEXKEL (0=not excluded/included; 1=excluded/not included)
#  ***What happens if there are no flags provided?***
n <- factor(flags$FLGEXKEL, levels=c(0,1))
ninc <- length(n[n==0])
nexc <- length(n[n==1])

#### Final changes to test ----------------------------------------------- ####
### Optimise Kel
# mct$OPTIMIZEKEL <- 0 # don't optimise
mct$OPTIMIZEKEL <- 1 # optimise

### Change Kel flags
flags$FLGEXKEL <- rep(1, nrow(flags))

#### Run computation engine ---------------------------------------------- ####
### Run computation
results_list <- run_computation(data = d, map = mct, flag = flags, 
  parameterset = parameterset)

### Save results to objects
r <- results_list$data_out
e <- results_list$est_data
f <- results_list$flag_data

### GitHub Output
head(r[c("KEL", "KELTMLO", "KELTMHI", "KELNOPT", "KELRSQ", "KELRSQA", "FLGACCEPTKEL")])
```

**Comment by jhhughes256** [comment]
### Update

As of ae10cfa1d9 this issue can still be replicated for tc011 and tc014.

**Comment by jhhughes256** [comment]
### Update

Labelling this as priority as it prevents the testing of OPTIMIZEKEL for tc011 and tc014 and has been brought up in issues testing FLGNOCMAX #21 .

**Comment by rudraya** [comment]
This issue was addressed in 02b2470, please let me know if it still occurs. I tested tc011 and tc014 with OPTMIZEKEL and FLGNOCMAX.

**Comment by jhhughes256** [comment]
### Update

This issue still occurs for me. I have noticed the following warning being given when running tc011 and tc014, which may be of help. 

```
In run_M2_SS_computation(data = merged_data, map = map_data, method = method,  :
  Kel optimization cannot be performed because 'TMAXi', 'TLASTi', 'CMAXi', 'CLASTi', 'AUCLASTi' are not part of the calculated parameters AND Flag 'FLGACCEPTKELCRIT' and Flag 'FLGXKEL' are not present in the dataset
```

Based on this warning I checked the mct and was able to find TMAXi, CMAXi, CLASTi & AUCLASTi in the mct, along with a valid FLGACCEPTKELCRIT value. The warning mentions FLGXKEL which I imagine is a typo.

**Comment by jhhughes256** [comment]
This issue also occurs when using the following changes (same changes as for tc011 and tc014) for tc071. It also provides the same warning.

```
mct$OPTIMIZEKEL <- 1
flags$FLGEXKEL <- rep(1, nrow(flags))
```

**Comment by rudraya** [comment]
This issue occurs because the "TLASTi" is not in PARAMETERLIST, and "TLASTi" and "CLASTi" is not in PARAMETERDISPLAYLIST. In order to be consistent with SS models, I added the case to verify if "TMAXi", 'CMAXi", "TLASSTi", "CLASTi" and "AUCLASTi" (which was previously "TMAX", 'CMAX", "TLASST", "CLAST" and "AUCLAST") are present in the parameterset in order to calculate OPTIMIZEKEL. And yes there is typo in the error that gets printed, it should say "FLGEXKEL" instead of "FLGXKEL".

**Comment by jhhughes256** [comment]
I can confirm that OPTIMIZEKEL works for tc011 and tc071 when adding "TLASTi" to `mct$PARAMETERLIST`. Thanks for the prompt response @rudraya . Am I correct in understanding that this is an issue with the input and not the computation engine?

**Comment by jhhughes256** [comment]
Just talked to @tensfeldt about this. TLASTi shouldn't be required to be added to the parameter_list simply to perform optimization. He has suggested using the create_dependency_list.R routine to resolve these kinds of computation dependencies and not require the user to add parameters that are not required for the analysis.

**Comment by rudraya** [comment]
This issue was addressed in 18821dc, please let me know if it still occurs.

**Comment by jhhughes256** [comment]
### Update

This issue no longer occurs as of 18821dc. As the issue outlining the error with OPTIMIZEKEL in tc014 is already documented (#58) this issue will be closed as it is issue #58 with OPTIMIZEKEL itself, rather than an issue with OPTIMIZEKEL not running.
